### ISSUE TITLE
Poetry applies python version of dependencies to project

### ISSUE THREAD

**Issue by cfra** [python-poetry/poetry#7510]
This is basically what is written here:

https://python-poetry.org/docs/faq/#why-is-poetry-telling-me-that-the-current-projects-python-requirement-is-not-compatible-with-one-or-more-packages-python-requirements

One example, with python stated as `^3.10` and installing `scipy`:

```console
$ poetry add scipy
Using version ^1.10.0 for scipy

Updating dependencies
Resolving dependencies... (0.0s)

The current project's Python requirement (>=3.10,<4.0) is not compatible with some of the required packages Python requirement:
  - scipy requires Python <3.12,>=3.8, so it will not be satisfied for Python >=3.12,<4.0

Because no versions of scipy match >1.10.0,<2.0.0
 and scipy (1.10.0) requires Python <3.12,>=3.8, scipy is forbidden.
So, because openai-playground depends on scipy (^1.10.0), version solving failed.

  • Check your dependencies Python requirement: The Python requirement can be specified via the `python` or `markers` properties
    
    For scipy, a possible solution would be to set the `python` property to ">=3.10,<3.12"

    https://python-poetry.org/docs/dependency-specification/#python-restricted-dependencies,
    https://python-poetry.org/docs/dependency-specification/#using-environment-markers
```

I am currently running python 3.10, so with current versions, this is just fine.

If I understand the message correctly, the issue is that if I had python 3.12, the situation would become unresolveable because scipy depends on `Python <3.12`.

Why is it a good idea to require the intersection of the possible pyton versions in the projects python version?

It feels a bit like saying: "You cannot depend on Django 4.0 because one of your dependencies depends on Django <= 3.2", which is obviously something the dependency solver should figure out.

Basically, my question is:

With dependencies, I specify which version _my_ code needs and the solver figures out if there is a solution that satisfies _my_ needs and the needs of _other_ packages.

Why should it be any different with the Python version I am depending on? Why is it a good idea to account for _other_ packages requirements in _my_ project?

**Comment by dimbleby** [comment]
do you want your project to support python 3.12 or not?

- if yes then scipy 1.10.0 rules itself out
- if not, then don't say `python = "^3.10"`

**Comment by cfra** [comment]
If I write `Django = ">=3.2,<4.3"` in my projects dependencies, it states that _my_ code works with any Django version from 3.2 to 4.2, but it doesn't say anything about whether the whole project supports Django 4.0 or Django 4.2.

It could very well be that Django 3.2 is the only valid solution because of a Django version constraints in some dependency.

Nobody would expect me to update my projects dependencies to `Django = "^3.2.0"` because some dependency warrants this.

Yet for the Python version Poetry requires this, and I don't fully understand the reasoning, especially because a requirement e.g. on `scipy = "^1.10"` could very well become solvable for Python 3.12 with a future scipy release.

I guess one reason why this is treated so differently could be that the Python version in the pyproject.toml doesn't follow the semantics of a dependency at all, although it is using the same syntax, but rather is a marker which Python versions the project is compatible with?

**Comment by dimbleby** [comment]
An installer doesn't get to resolve a python version, which makes python different from all other dependencies.

eg if someone running python 3.12 tries to install a project that claims compatibility with `python = "^3.10"`, then they should expect that to succeed.
